hazelcast/src/main/java/com/hazelcast/sql/impl/expression/datetime/Formatter.java

```diff
@@ -0,0 +1,791 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression.datetime;
+
+import com.hazelcast.sql.impl.QueryException;
+
+import java.io.Serializable;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.math.MathContext;
+import java.math.RoundingMode;
+import java.text.DecimalFormat;
+import java.text.DecimalFormatSymbols;
+
+import java.time.DayOfWeek;
+import java.time.ZoneOffset;
+import java.time.format.DateTimeFormatter;
+import java.time.format.TextStyle;
+import java.time.temporal.ChronoField;
+import java.time.temporal.IsoFields;
+import java.time.temporal.JulianFields;
+import java.time.temporal.Temporal;
+import java.time.temporal.TemporalField;
+import static java.time.temporal.WeekFields.ISO;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.function.BiFunction;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Implementation of <a href="https://www.postgresql.org/docs/14/functions-formatting.html">
+ * PostgreSQL <code>to_char()</code></a> with the following differences.
+ * <ol><li> {@code FM} and {@code TM} prefixes enable the fill and translation modes for date
+ *          patterns respectively, which are disabled by default.
+ *     <li> {@code V} pattern can be combined with a decimal point.
+ *     <li> {@code M} and {@code P} patterns are introduced as the anchored versions of {@code MI}
+ *          and {@code PL} patterns respectively.
+ *     <li> {@code PR} pattern is replaced with {@code B} and {@code BR} patterns. The former is
+ *          the anchored version of the latter.
+ *     <li> {@code L} pattern is replaced with {@code C} and {@code CR} patterns. The former is the
+ *          anchored version of the latter.
+ *     <li> Literals are anchored by default. To fix its position, a literal should be prepended
+ *          with an {@code F} pattern.
+ *     <li> Zero-padding and space-padding are completely orthogonal, which makes it possible to
+ *          have zero-padded fractions, which are aligned at the the decimal separator. However,
+ *          this necessitates to have a {@code '0'} at the end of the pattern if the PostgreSQL
+ *          convention is desired. </ol>
+ */
+@SuppressWarnings({"checkstyle:CyclomaticComplexity", "checkstyle:DeclarationOrder",
+        "checkstyle:ExecutableStatementCount", "checkstyle:InnerAssignment", "checkstyle:MagicNumber",
+        "checkstyle:MethodLength", "checkstyle:MultipleVariableDeclarations", "checkstyle:NestedIfDepth",
+        "checkstyle:NPathComplexity"})
+public class Formatter implements Serializable {
```

Comment:
Should I make `Formatter` an `IdentifiedDataSerializable` and add it to `SqlDataSerializerHook`, or whitelist it for `DataSerializableConventionsTest.test_serializableClasses_areIdentifiedDataSerializable()`?

Reply:
It would be better to convert to IDS, but it will be quite a lot of boilerplate, so let's add an exception for now.

Reply:
After checking the code more, I realized the Formatter sent over the wire only has the `format` field non-null. This is done for the constant pattern. But there's no point in doing it this way. The point is that we cache the parsed format in the plan cache. If we're sending just the string and parsing the format on first use, then we should send just `String`, and not `Formatter`. I realized this because I noticed we miss `Serializable` interface at `Literal`, but an integration test with a literal didn't fail.

The problem is even worse: due to the lazy initialization the Formatter is mutable, and it can be accessed concurrently. It would be better if it's immutable.

So we should do either:
- remove the `Formatter.format` field and add a test that all possible Formatter instances serialize (perhaps in the FormatterTest try to ser-de all the instances.
- send just the string and remove `Serializable` from `Formatter`.

My gut feeling is that parsing might be faster than java deserialization.